### PR TITLE
ARROW-14906: [C++] Enable CSV Writer to control the type of escape used for quoting

### PR DESCRIPTION
**Pull request by ahadnagy**

### CODE REVIEW

File: cpp/src/arrow/csv/writer.cc [hunk 1]
```diff
@@ -126,13 +126,14 @@ class ColumnPopulator {
...
 // Copies the contents of to out properly escaping any necessary characters.
 // Returns the position prior to last copied character (out_end is decremented).
-char* EscapeReverse(arrow::util::string_view s, char* out_end) {
+char* EscapeReverse(arrow::util::string_view s, char* out_end, bool escaping,
+                    char escape_char) {
   for (const char* val = s.data() + s.length() - 1; val >= s.data(); val--, out_end--) {
-    if (*val == '"') {
-      *out_end = *val;
+    *out_end = *val;
+    if (escaping && *val == '"') {
```
Comment (lidavidm) [inline cpp/src/arrow/csv/writer.cc:133]: Shouldn't we also test for the escape char?

Comment (lidavidm) [reply cpp/src/arrow/csv/writer.cc:133]: Also CountEscapes (or CountQuotes) will also need to test for the escape char.

File: cpp/src/arrow/csv/writer.cc [hunk 2]
```diff
@@ -184,9 +185,11 @@ class UnquotedColumnPopulator : public ColumnPopulator {
 // a quote character (") and escaping is done my adding another quote.
 class QuotedColumnPopulator : public ColumnPopulator {
```
Comment (lidavidm) [inline cpp/src/arrow/csv/writer.cc:186]: Note the above docstring is outdated now.

File: cpp/src/arrow/csv/writer.cc (hunk 1 above)
Comment (lidavidm) [reply cpp/src/arrow/csv/writer.cc:133]: (Though, once merged with #11849: that PR uses CountQuotes to check that quotes don't appear when quoting is disabled, however, we need to make sure that doesn't cause spurious errors if the escape character appears in an unquoted cell. So looks like we need both CountQuotes and CountEscapes.)

Comment (ahadnagy) [reply cpp/src/arrow/csv/writer.cc:133]: https://www.rdocumentation.org/packages/readr/versions/1.3.1/topics/write_delim
If I understand correctly, the R function we're trying to align with only escapes quotes.

(Correspond line in the Vroom source: https://github.com/r-lib/vroom/blob/c5b115b1b5f21852886d6407317386b215706239/src/vroom_write.cc#L146)

Comment (lidavidm) [reply cpp/src/arrow/csv/writer.cc:133]: Ah, I see. Then that simplifies things greatly.

File: cpp/src/arrow/csv/writer.cc [hunk 3]
```diff
@@ -195,11 +199,17 @@ class QuotedColumnPopulator : public ColumnPopulator {
     VisitArrayDataInline<StringType>(
         *input.data(),
         [&](arrow::util::string_view s) {
-          int64_t escaped_count = CountEscapes(s);
-          // TODO: Maybe use 64 bit row lengths or safe cast?
-          row_needs_escaping_[row_number] = escaped_count > 0;
-          row_lengths[row_number] += static_cast<int32_t>(s.length()) +
-                                     static_cast<int32_t>(escaped_count + kQuoteCount);
+          if (escaping_) {
+            int64_t escaped_count = CountEscapes(s);
```
Comment (lidavidm) [inline cpp/src/arrow/csv/writer.cc:203]: If escaping is disabled, perhaps we should still count escapes and error if the number is non-zero?

File: cpp/src/arrow/csv/writer.cc [hunk 4]
```diff
@@ -390,7 +407,7 @@ class CSVWriterImpl : public ipc::RecordBatchWriter {
     for (int col = 0; col < schema_->num_fields(); col++) {
       const std::string& col_name = schema_->field(col)->name();
       header_length += col_name.size();
-      header_length += CountEscapes(col_name);
+      header_length += options_.escaping ? CountEscapes(col_name) : 0;
```
Comment (lidavidm) [inline cpp/src/arrow/csv/writer.cc:410]: The same goes here, if escaping is disabled, we should ensure that no escaping is actually required.

Comment (lidavidm) [reply cpp/src/arrow/csv/writer.cc:410]: (We should also cover these with tests.)

**Comment by pitrou** [comment]
Hmm, wouldn't it be better to wait for PR #11849 before continuing work on this one?

File: cpp/src/arrow/csv/writer.cc [hunk 5]
```diff
@@ -253,6 +263,8 @@ class QuotedColumnPopulator : public ColumnPopulator {
   // at some point we should change this to use memory_pool
   // backed allocator.
   std::vector<bool> row_needs_escaping_;
+  bool escaping_;
```
Comment (emkornfield) [inline cpp/src/arrow/csv/writer.cc:266]: nit: const for this and escape_char_

File: cpp/src/arrow/csv/writer_test.cc [hunk 6]
```diff
@@ -49,6 +49,8 @@ struct WriterTestParams {
   WriteOptions options;
   util::optional<std::string> expected_output;
   Status expected_status;
+  bool options_invalid;
+  bool escaping_invalid;
```
Comment (lidavidm) [inline cpp/src/arrow/csv/writer_test.cc:53]: It seems these aren't used anymore?

File: cpp/src/arrow/csv/writer.cc [hunk 7]
```diff
@@ -462,25 +483,32 @@ class CSVWriterImpl : public ipc::RecordBatchWriter {
...
+      header_length += escaped_count;
     }
     return header_length + (kQuoteDelimiterCount * schema_->num_fields());
   }
 
   Status WriteHeader() {
     // Only called once, as part of initialization
-    RETURN_NOT_OK(data_buffer_->Resize(CalculateHeaderSize(), /*shrink_to_fit=*/false));
+    int64_t header_size;
+    ASSIGN_OR_RAISE(header_size, CalculateHeaderSize());
```
Comment (lidavidm) [inline cpp/src/arrow/csv/writer.cc:503]: N.B. the ASSIGN_OR_RAISE macro lets you declare the variable inline (`ASSIGN_OR_RAISE(int64_t header_size, CalculateHeaderSize());`)

**Review by lidavidm — COMMENTED** [review]
Thanks! Just one nit.

We should also file a follow-up to expose all the new options in Python.

**Comment by pitrou** [comment]
1) Can you rebase this and fix conflicts?
2) It seems it would be nice to have a common setting for quoting and escaping (either you quote or you escape special chars, not both at once)

**Comment by ahadnagy** [comment]
> 2. It seems it would be nice to have a common setting for quoting and escaping (either you quote or you escape special chars, not both at once)

These are different options in both Pandas (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html) and readr (https://readr.tidyverse.org/reference/write_delim.html). I'd lean towards aligning with these, but if you have a proposal, we can discuss that.

**Comment by pitrou** [comment]
@ahadnagy If you have both quoting and escaping enabled, how do you decide whether you should quote or escape a special character?

**Comment by ahadnagy** [comment]
Currently, the escape character is being used (this aligns with readr).
The Pandas csv_writer has a `doublequote` parameter to control quoting/escaping inside a field explicitly.

**Comment by pitrou** [comment]
Right, but this also makes the whole API confusing (why is the escape character being used? what if both double-quoting and escaping are enabled?).

**Comment by pitrou** [comment]
Let me suggest the following API:
```c++
  /// \brief Quoting style
  QuotingStyle quoting_style = QuotingStyle::Needed;

  /// Whether a quote inside a quoted value is double-quoted (if true)
  /// or escaped using the `escape_char` (if false).
  bool double_quote = true;

  /// Escaping character for non-quoted fields.
  char escape_char = kDefaultEscapeChar;
```

with the following semantics:
* if in a quoted field and `double_quote` is true, a quote in a value is double-quoted (i.e. becomes `""`)
* if in a quoted field and `double_quote` is false, a quote in a value is escape (e.g. becomes `\"` for the default escape char)
* if in a non-quoted field, any special character (quote, field delimiter, end of line character) is escaped using the escape char

**Comment by amol-** [comment]
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍

**Pull request closed without merging**